ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -24,90 +27,251 @@
+
 
 /**
- * BlobId uniquely identifies a stored blob as well as the Partition in which the blob is stored.
+ * <p>
+ *   BlobId primarily consists of a uuid to uniquely identifies a stored blob. A blobId is a reference that is
+ *   returned back to a caller when posting a blob, and later will be required to fetch the blob.
+ * </p>
+ * <p>
+ *   There are two versions of format for blob de/serialization. Version 1, which includes {@code partitionId}
+ *   of the blob. The {@code partitionId} is the {@link com.github.ambry.clustermap.Partition} to which this
+ *   blob is assigned.
+ * </p>
+ * <pre>
+ * +---------------------------------------------+
+ * | version | partitionId | uuidSize | uuid     |
+ * | (short) | (n bytes)   | (int)    | (n bytes)|
+ * +---------------------------------------------+
+ * </pre>
+ * <p>
+ *   Version 2, which includes {@code flag}, {@code accountId}, {@code containerId}, {@code datacenterId},
+ *   and {@code partitionId} of the blob. {@code flag} is a single byte that carries the meta information
+ *   for this blobId. It will be a place holder and assigned {@link #DEFAULT_FLAG} (i.e., no bit is set)
+ *   before the assignment of each bit is determined. The {@code datacenterId} is the id of the datacenter
+ *   where this blob was originally posted (not through replication). The {@code accountId} is the id of
+ *   the {@link Account} the blob belongs to. The {@code containerId} is the {@link Container} the blob
+ *   belongs to. The {@code partitionId} is the {@link com.github.ambry.clustermap.Partition} to which this
+ *   blob is assigned.
+ * </p>
+ * <pre>
+ * +---------+-------+--------------+-----------+-------------+-------------+----------+----------+
+ * | version | flag  | datacenterId | accountId | containerId | partitionId | uuidSize | uuid     |
+ * | (short) | (byte)| (byte)       | (short)   | (short)     | (n bytes)   | (int)    | (n bytes)|
+ * +---------+----------------------+-----------+-------------+-------------+----------+----------+
+ * </pre>
  */
 public class BlobId extends StoreKey {
-  private Short version = 1;
-  private static short Version_Size_In_Bytes = 2;
+  // version 1 of the serialized format
+  static final short BLOB_ID_V1 = 1;
+  // version 2 of the serialized format
+  static final short BLOB_ID_V2 = 2;
+  static final byte DEFAULT_FLAG = 0;
+  private static final short CURRENT_VERSION = BLOB_ID_V1;
+  private static final short VERSION_FIELD_LENGTH_IN_BYTES = 2;
+  private static final short UUID_SIZE_FIELD_LENGTH_IN_BYTES = 4;
+  private static final short FLAG_FIELD_LENGTH_IN_BYTES = 1;
+  private static final short DATACENTER_ID_FIELD_LENGTH_IN_BYTES = 1;
+  private static final short ACCOUNT_ID_FIELD_LENGTH_IN_BYTES = 2;
+  private static final short CONTAINER_ID_FIELD_LENGTH_IN_BYTES = 2;
+  // the version to indicate the serialized format.
+  private Short version;
+  private Byte flag;
+  private Byte datacenterId;
```

Comment:
They cannot be declared as final because we have the logic to populate missing fields.